### PR TITLE
Picker: bindable SelectedItem & ItemsSource props, Entry: IsFocused prop not only ReadOnly

### PR DESCRIPTION
**Pull request by samih7**
### Description of Change ###
Picker.cs:
- A Collection can be bound to the ItemsSource property.
- We can easily get the current selected item with the SelectedItem property.

Entry.cs:
- An Entry can now be focused (call to the Focus(); method in the propertyChanged delegate) with the 'IsFocused' property. In order to do that, I have hidden 'IsFocused' that is inherited from VisualElement (new keyword).

### Bugs Fixed ###
None

### API Changes ###
None

Added:
Picker.cs:
- public static readonly BindableProperty ItemsSourceProperty = BindableProperty.Create (nameof (ItemsSource), typeof (IEnumerable), typeof (Picker), null, propertyChanged: OnItemsSourcePropertyChanged);

- public static readonly BindableProperty SelectedItemProperty = BindableProperty.Create (nameof (SelectedItem), typeof (object), typeof (Picker), null, BindingMode.TwoWay, propertyChanged: OnSelectedItemPropertyChanged);

- static void OnItemsSourcePropertyChanged (BindableObject bindable, object value, object newValue);

- static void OnSelectedItemPropertyChanged (BindableObject bindable, object value, object newValue);

Entry.cs:
- public static new readonly BindableProperty IsFocusedProperty = BindableProperty.Create (nameof (IsFocused), typeof (bool), typeof (Entry), false, BindingMode.TwoWay, propertyChanged: OnIsFocusedChanged);

- public new bool IsFocused;

- public static void OnIsFocusedChanged (object sender, object oldValue, object newvalue);
- public static void OnIsFocusedChanged (VisualElement sender, bool oldValue, bool newValue);

Changed:
 - public Picker(); => public Picker(); (added : SelectedIndexChanged += (o, e) => ...)

### Behavioral Changes ###

Picker.cs:
- Two bindable properties added.

Entry.cs:
- One bindable property added.

### PR Checklist ###

- [ ] Has tests (if omitted, state reason in description)
- [x] Rebased on top of master at time of PR
- [x] Changes adhere to coding standard
- [ ] Consolidate commits as makes sense

### CODE REVIEW

**Comment by ahmedalejo** [comment]
Should unsubscribe from oldValue's CollectionChanged event if is
INotifyCollectionChanged

**Comment by ahmedalejo** [comment]
That's in the ItemSourceChanged handler

File: Xamarin.Forms.Core/Entry.cs [hunk 1]
```diff
@@ -95,5 +103,17 @@ static void OnTextChanged(BindableObject bindable, object oldValue, object newVa
...
+		{
+			OnIsFocusedChanged ((VisualElement)sender, (bool)oldValue, (bool)newvalue);
+		}
+
+		public static void OnIsFocusedChanged (VisualElement sender, bool oldValue, bool newValue)
+		{
+			if (newValue)
+				sender.Focus ();
+			sender.SetValue (IsFocusedProperty, false);
+		}
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/Entry.cs]: Could you move those Entry changes to a different PR?
This is a huge change, it certainly requires changes on each platforms. It'd require tons of UI and unit tests to make sure that
- only one element is focused (i.e. has the IsFocused property) at any time
- it works well in even the more complex solutions

File: Xamarin.Forms.Core/Picker.cs [hunk 2]
```diff
@@ -1,6 +1,8 @@
 using System;
+using System.Collections;
 using System.Collections.Generic;
 using System.Collections.Specialized;
+using System.Linq;
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/Picker.cs]: We try to avoid using Linq extension methods, most of them are generating too many unrequited allocations

File: Xamarin.Forms.Core/Picker.cs [hunk 3]
```diff
@@ -46,6 +56,16 @@ public string Title
 			set { SetValue(TitleProperty, value); }
 		}
 
+		public IEnumerable<object> ItemsSource {
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/Picker.cs]: Why is the BP `ItemsSourceProperty` an `IEnumerable` and the property an `IEnumerable<object>` ?

File: Xamarin.Forms.Core/Picker.cs [hunk 4]
```diff
@@ -58,5 +78,39 @@ void OnItemsCollectionChanged(object sender, NotifyCollectionChangedEventArgs e)
 		{
 			SelectedIndex = SelectedIndex.Clamp(-1, Items.Count - 1);
 		}
+
+		static void OnItemsSourcePropertyChanged (BindableObject bindable, object value, object newValue)
+		{
+			var picker = (Picker)bindable;
+			var notifyCollection = newValue as INotifyCollectionChanged;
+			if (notifyCollection != null) {
+				notifyCollection.CollectionChanged += (sender, args) => {
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/Picker.cs]: This event handler is never unhooked, and it will leak

File: Xamarin.Forms.Core/Picker.cs [hunk 5]
```diff
@@ -58,5 +78,39 @@ void OnItemsCollectionChanged(object sender, NotifyCollectionChangedEventArgs e)
...
+				picker.Items.Add ((item ?? "").ToString ());
+		}
+
+		static void OnSelectedItemPropertyChanged (BindableObject bindable, object value, object newValue)
+		{
+			var picker = (Picker)bindable;
+			if (picker.ItemsSource != null)
+				picker.SelectedIndex = picker.ItemsSource.IndexOf(picker.SelectedItem);
+		}
+
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/Picker.cs]: What will happen if someone manually changes the Items ? Or the SelectedIndex ?

File: Xamarin.Forms.Core/Picker.cs [hunk 6]
```diff
@@ -58,5 +78,39 @@ void OnItemsCollectionChanged(object sender, NotifyCollectionChangedEventArgs e)
...
+				notifyCollection.CollectionChanged += (sender, args) => {
+					if (args.NewItems != null)
+						foreach (var newItem in args.NewItems)
+							picker.Items.Add ((newItem ?? "").ToString ());
+
+					if (args.OldItems != null)
+						foreach (var oldItem in args.OldItems)
+							picker.Items.Remove ((oldItem ?? "").ToString ());
+				};
+			}
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/Picker.cs]: As elements are object in the ItemsSource, you want to listen to INPC on each of them, and update the list accordingly

File: Xamarin.Forms.Core/Picker.cs [hunk 7]
```diff
@@ -58,5 +78,39 @@ void OnItemsCollectionChanged(object sender, NotifyCollectionChangedEventArgs e)
...
+
+			var enumerable = newValue as IEnumerable;
+			if (enumerable == null)
+				return;
+
+			picker.Items.Clear ();
+
+			foreach (var item in enumerable)
+				picker.Items.Add ((item ?? "").ToString ());
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/Picker.cs]: You mean `item?.ToString() ?? ""` ??

Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/Picker.cs]: Using ToString is probably not what we want here. A mapping function, or a MemberPath should be preferred. And if we do this, there's another control (can't recall which one right now) in Forms that needs the exact same treatment.

**Comment by StephaneDelcroix** [comment]
Thanks @samih7 for this PR. We understand the use case and the need for this enhancement. Unfortunately, the proposed solution isn't quite there yet. Hence my 👎
